quickwit/quickwit-proto/src/lib.rs

```diff
@@ -297,11 +296,60 @@ pub fn set_parent_span_from_request_metadata(request_metadata: &tonic::metadata:
     Span::current().set_parent(parent_cx);
 }
 
+/// Index identifiers that uniquely identify not only the index, but also
+/// its incarnation allowing to distinguish between deleted and recreated indexes.
+/// It is represented as a stiring in index_id:incarnation_id format.
+#[derive(Clone, Debug, Default, Serialize, Deserialize, PartialEq, Eq, Ord, PartialOrd, Hash)]
+pub struct IndexUid(String);
+
+impl IndexUid {
+    /// Creates a new index uid form index_id and incarnation_id
+    pub fn new(index_id: impl ToString) -> Self {
+        Self::from_parts(index_id,  Ulid::new().to_string())
+    }
+
+    pub fn from_parts(index_id: impl ToString, incarnation_id: impl ToString) -> Self {
+        Self (
+            format!("{}:{}", index_id.to_string(), incarnation_id.to_string())
+        )
+    }
+
+    pub fn index_id(&self) -> &str {
+        self.0.split(':').next().unwrap()
+    }
+
+    pub fn incarnation_id(&self) -> &str {
+        self.0.rsplit(':').next().unwrap()
+    }
+
+    pub fn is_empty(&self) -> bool {
+        self.0.is_empty()
+    }
+}
+
+impl From<IndexUid> for String {
+    fn from(val: IndexUid) -> Self {
+        val.0
+    }
+}
+
+impl ToString for IndexUid {
+    fn to_string(&self) -> String {
+        self.0.clone()
+    }
+}
+
+impl From<String> for IndexUid {
```

Comment:
It is a conversion from `String` to `IndexUid(String)`. I don't think it is more dangerous than passing index name in String, which it essentially replaces. It is also used only for internal deserialization from our messages and not something provided by the end-user. It can be abused but only if somebody is messing with messages on gprs level. If we want to protect that level from fuzzing, it would require a much bigger effort.